util/noeintr_close_shared.c

```diff
@@ -0,0 +1,212 @@
+#include <sys/socket.h>
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "entropy.h"
+#include "warnp.h"
+
+#include "noeintr.h"
+
+/* Sanity check compiling. */
+#ifndef NOEINTR_CLOSE_PTHREAD
+#error "Do not use this file directly; compile _normal.c or _pthread.c instead"
+#endif
+
+/* Handle multithreading (if applicable). */
+#if NOEINTR_CLOSE_PTHREAD
+#include <pthread.h>
+
+static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+#endif
+
+#define MAGIC_COOKIE_LEN 32
+
+static int initialized = 0;
+static int s[2] = {-1, -1};
+static uint8_t cookie_write[MAGIC_COOKIE_LEN];
+
+#ifdef NOEINTR_CLOSE_TESTING
+/*
+ * POSIX is silent about the state of fd if close(2) gets EINTR.  For internal
+ * testing, this function randomly selects between the return values which are
+ * permitted by the standard.  For more info, see:
+ * https://www.daemonology.net/blog/2011-12-17-POSIX-close-is-broken.html
+ * https://www.austingroupbugs.net/view.php?id=529
+ */
+static int
+evil_close(int fd)
+{
+
+	switch (random() % 3) {
+	case 0:
+		/* Normal close(). */
+		return (close(fd));
+	case 1:
+		/* No close(), then return EINTR. */
```

Comment:
`Fail with EINTR without changing the open file description to which fd refers.`